src/ServiceManager.php

```diff
@@ -578,55 +581,37 @@ private function createDelegatorFromName($name, ?array $options = null)
         foreach ($this->delegators[$name] as $index => $delegatorFactory) {
             $delegatorFactory = $this->delegators[$name][$index];
 
-            if ($delegatorFactory === Proxy\LazyServiceFactory::class) {
+            if ($delegatorFactory === LazyServiceFactory::class) {
                 $delegatorFactory = $this->createLazyServiceDelegatorFactory();
-            }
-
-            if (is_string($delegatorFactory) && class_exists($delegatorFactory)) {
+            } elseif (is_string($delegatorFactory) && class_exists($delegatorFactory)) {
                 $delegatorFactory = new $delegatorFactory();
             }
 
-            if (! is_callable($delegatorFactory)) {
-                if (is_string($delegatorFactory)) {
-                    throw new ServiceNotCreatedException(sprintf(
-                        'An invalid delegator factory was registered; resolved to class or function "%s"'
-                        . ' which does not exist; please provide a valid function name or class name resolving'
-                        . ' to an implementation of %s',
-                        $delegatorFactory,
-                        DelegatorFactoryInterface::class
-                    ));
-                }
-
-                throw new ServiceNotCreatedException(sprintf(
-                    'A non-callable delegator, "%s", was provided; expected a callable or instance of "%s"',
-                    is_object($delegatorFactory) ? get_class($delegatorFactory) : gettype($delegatorFactory),
-                    DelegatorFactoryInterface::class
-                ));
-            }
+            $this->assertCallableDelegatorFactory($delegatorFactory);
 
+            /**
+             * @psalm-var callable(ContainerInterface,string,callable,array<mixed>|null):object $delegatorFactory
+             */
             $this->delegators[$name][$index] = $delegatorFactory;
 
             $creationCallback = function () use ($delegatorFactory, $name, $creationCallback, $options) {
                 return $delegatorFactory($this->creationContext, $name, $creationCallback, $options);
             };
         }
 
-        return $creationCallback($this->creationContext, $name, $creationCallback, $options);
+        return $creationCallback();
     }
 
     /**
      * Create a new instance with an already resolved name
      *
      * This is a highly performance sensitive method, do not modify if you have not benchmarked it carefully
      *
-     * @param  string     $resolvedName
-     * @param  null|array $options
-     * @return mixed
      * @throws ServiceNotFoundException If unable to resolve the service.
      * @throws ServiceNotCreatedException If an exception is raised when creating a service.
      * @throws ContainerException If any other error occurs.
      */
-    private function doCreate($resolvedName, ?array $options = null)
+    private function doCreate(string $resolvedName, ?array $options = null): object
```

Comment:
This will break with `array` services, due to the newly declared `: object` return value

Reply:
`array` services are not retrievable via `doCreate` as this method only uses `factories` or `abstract_factories` which both must return `object` as per definition.

`array` is "usually" only `ServiceManager#setService('config', $configArray);` or somehow else but it actually is not possible to create an array via a factory.

anyways, I've converted this to a soft type-hint